File: smallpict-java/src/main/java/com/smallpict/errors/SmallPictException.java
package com.smallpict.errors;

import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class SmallPictException extends RuntimeException {
    private static final Pattern KEY_PATTERN = Pattern.compile("sp_(live|test|sdk|wp)_[a-zA-Z0-9_-]{10,}");
    private static final Pattern SECRET_PATTERN = Pattern.compile("(?i)(sec|secret)_[a-zA-Z0-9_-]{8,}");
    private static final Pattern BEARER_PATTERN = Pattern.compile("(?i)Bearer\\s+[a-zA-Z0-9._-]+");

    private final String errorCode;
    private final Integer statusCode;
    private final String requestId;
    private final Map<String, Object> details;

    public SmallPictException(String message) {
        this(message, "INTERNAL_ERROR", null, null, null, null);
    }

    public SmallPictException(
            String message,
            String errorCode,
            Integer statusCode,
            String requestId,
            Map<String, Object> details,
            Throwable cause) {
        super(sanitize(message), cause);
        this.errorCode = errorCode != null ? errorCode : "INTERNAL_ERROR";
        this.statusCode = statusCode;
        this.requestId = requestId;
        this.details = details;
    }

    public static String sanitize(String msg) {
        if (msg == null || msg.isEmpty()) {
            return "";
        }

        Matcher keyMatcher = KEY_PATTERN.matcher(msg);
        StringBuffer sb = new StringBuffer();
        while (keyMatcher.find()) {
            String full = keyMatcher.group();
            String masked = full.length() > 14
                    ? full.substring(0, 10) + "..." + full.substring(full.length() - 4)
                    : full.substring(0, 6) + "...";
            keyMatcher.appendReplacement(sb, Matcher.quoteReplacement(masked));
        }
        keyMatcher.appendTail(sb);
        String step1 = sb.toString();

        String step2 = SECRET_PATTERN.matcher(step1).replaceAll("***REDACTED***");
        return BEARER_PATTERN.matcher(step2).replaceAll("Bearer ***REDACTED***");
    }

    public String getErrorCode() {
        return errorCode;
    }

    public Integer getStatusCode() {
        return statusCode;
    }

    public String getRequestId() {
        return requestId;
    }

    public Map<String, Object> getDetails() {
        return details;
    }

    @Override
    public String toString() {
        String statusStr = statusCode != null ? " HTTP " + statusCode : "";
        String reqStr = requestId != null ? " (Request ID: " + requestId + ")" : "";
        return "[" + errorCode + statusStr + "]: " + getMessage() + reqStr;
    }
}

File: smallpict-php/src/Exceptions/SmallPictException.php
<?php

declare(strict_types=1);

namespace SmallPict\Exceptions;

class SmallPictException extends \Exception
{
    protected string $errorCode;
    protected ?int $statusCode;
    protected ?string $requestId;
    /**
     * @var array<string, mixed>
     */
    protected array $details;

    /**
     * @param string $message
     * @param string $errorCode
     * @param int|null $statusCode
     * @param string|null $requestId
     * @param array<string, mixed> $details
     * @param \Throwable|null $previous
     */
    public function __construct(
        string $message = '',
        string $errorCode = 'INTERNAL_ERROR',
        ?int $statusCode = null,
        ?string $requestId = null,
        array $details = [],
        ?\Throwable $previous = null
    ) {
        $safeMessage = self::sanitizeMessage($message);
        parent::__construct($safeMessage, $statusCode ?? 0, $previous);

        $this->errorCode = $errorCode;
        $this->statusCode = $statusCode;
        $this->requestId = $requestId;
        $this->details = $details;
    }

    public static function sanitizeMessage(string $msg): string
    {
        if ($msg === '') {
            return $msg;
        }

        // Mask sp_live_..., sp_test_..., sp_sdk_..., sp_wp_...
        $msg = (string)preg_replace_callback(
            '/sp_(live|test|sdk|wp)_[a-zA-Z0-9_-]{10,}/',
            static function ($matches) {
                $full = $matches[0];
                return substr($full, 0, 10) . '...' . substr($full, -4);
            },
            $msg
        );

        // Mask secrets
        $msg = (string)preg_replace('/(sec|secret)_[a-zA-Z0-9_-]{8,}/i', '***REDACTED***', $msg);
        $msg = (string)preg_replace('/Bearer\s+[a-zA-Z0-9._-]+/i', 'Bearer ***REDACTED***', $msg);

        return $msg;
    }

    public function getErrorCode(): string
    {
        return $this->errorCode;
    }

    public function getStatusCode(): ?int
    {
        return $this->statusCode;
    }

    public function getRequestId(): ?string
    {
        return $this->requestId;
    }

    /**
     * @return array<string, mixed>
     */
    public function getDetails(): array
    {
        return $this->details;
    }

    public function __toString(): string
    {
        $statusStr = $this->statusCode ? " HTTP {$this->statusCode}" : '';
        $reqStr = $this->requestId ? " (Request ID: {$this->requestId})" : '';
        return "[SmallPictException] ({$this->errorCode}{$statusStr}): {$this->getMessage()}{$reqStr}";
    }
}
